sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/model/hdt/TripleOps.scala

```diff
@@ -0,0 +1,194 @@
+    * @param subjectDictDF Dictionary Dataframe of Subject
+    * @param objectDictDF Dictionary Dataframe of Object
+    * @param predicateDict Dictionary Dataframe of Predicate
+    */
+  def registerDictionariesAsView(subjectDictDF : DataFrame, objectDictDF : DataFrame, predicateDict : DataFrame) : Unit = {
+    subjectDictDF.createOrReplaceTempView(TripleOps.SUBJECT_TABLE)
+    objectDictDF.createOrReplaceTempView(TripleOps.OBJECT_TABLE)
+    predicateDict.createOrReplaceTempView(TripleOps.PREDICATE_TABLE)
+  }
+
+
+  /**
+    * Return Dataframe of Index + Subject by retrieving the unique subjects from RDD[Triple] and zip it with undex
+    * @param triples RDD[Triple] conversion of input file
+    * @return DataFrame Subject dictionary of [index,subject]
+    */
+  def getDistinctSubjectDictDF(triples : RDD[graph.Triple]) : DataFrame = {
+    spark.createDataFrame(triples.map(_.getSubject.toString()).distinct().zipWithIndex().map(t => Row(t._1, t._2)), dictionarySchema).cache()
+  }
+
+  /**
+    * Return Dataframe of Index + Predicate by retrieving the unique predicate from RDD[Triple] and zip it with undex
+    * @param triples RDD[Triple] conversion of input file
+    * @return DataFrame Predicate dictionary of [index,Prediate]
+    */
+  def getDistinctPredicateDictDF(triples : RDD[graph.Triple]) : DataFrame = {
+    spark.createDataFrame(triples.map(_.getPredicate.toString()).distinct().zipWithIndex().map(t => Row(t._1, t._2)), dictionarySchema).cache()
+  }
+
+  /**
+    * Return Dataframe of Index + Object by retrieving the unique objects from RDD[Triple] and zip it with undex
+    * @param triples RDD[Triple] conversion of input file
+    * @return DataFrame Object dictionary of [index , object]
+    */
+  def getDistinctObjectDictDF(triples : RDD[graph.Triple]) : DataFrame = {
+    spark.createDataFrame( triples.map(_.getObject.toString()).distinct().zipWithIndex().map(t => Row(t._1, t._2)), dictionarySchema).cache()
+  }
+
+
+  /**
+    * Reads the input rdf file and returns spark RDD[Triple]
+    * @param input Input RDF file path
+    * @return returns RDD[Tripe]
+    */
+  def readRDFFromFile( input : String) : RDD[graph.Triple] = {
+    val triples : RDD[graph.Triple] = spark.rdf(Lang.NTRIPLES)(input)
+    triples
+  }
+
+
+  /**
+    * This is key function of TripleOps that read RDF file and create Dictionaries and Index Table and register them as Spark In memory Table
+    * @param input Input RDF File Path [Either One of the input is require]
+    * @param compressedDir Input compressed-directory Path to read compressed data directly [Either One of the input is require]
+    * @param registerAsTable If true, it register all the DF as Spark table
+    * @return Returns the Tuple4 [IndexDataFrame,SubjectDictDataFrame,ObjectDictDataFrame,PredicateDictDataFrame]
+    */
+  def createOrReadDataSet(input : String, compressedDir : String, registerAsTable : Boolean = true) : (DataFrame, DataFrame, DataFrame, DataFrame) = {
```

Comment:
This method requires to have the compressed data already kept in the dist (which usually is mean to be a pre-processing step), did you also consider that sometimes there is no need to materialize the process and keep them in memory and do not put them back to the dist.  

I think we should provide another method which allows the users to read the data and compress them on the fly and or course to save them in case they want to do so.

Reply:
"createOrReadDataSet" method does this. If NTriple file is passed, it processes and converts it into index dictionary. If compressed Directory (i.e. already parsed) is passed, then it simply read it as indexed dataFrames. However, I will separate out the function into 2, one for reading from raw NT file and the second one from processed data.

There is already a provision of saving processed (compressed) dataset into a disk using saveAsCSV() and re-read.

Reply:
Hi @abakarboubaa ,

indeed. There is no need to do such a separate. I already did it here:
- [getHDT](https://github.com/SANSA-Stack/SANSA-RDF/blob/develop/sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/model/hdt/TripleOps.scala#L80) and [readHDTFromDisk](https://github.com/SANSA-Stack/SANSA-RDF/blob/develop/sansa-rdf-spark/src/main/scala/net/sansa_stack/rdf/spark/model/hdt/TripleOps.scala#L110)

See you